tests/test_initial_data_preparation.py

```diff
@@ -0,0 +1,321 @@
+
+
+@pytest.mark.parametrize(
+    "url, duplicate_flag, unsupported_url_scheme, expected_flag",
+    [
+        # Properly formatted URL with a query string, follows the complete
+        # structure including parameters. The presence of a query does not
+        # affect completeness.
+        ("https://example.com/over/there?name=ferret", False, False, False),
+        # URL contains special characters but still has the necessary parts.
+        ("http://example.com/owner/special'chars", False, False, False),
+        ("http://example.com/owner/repo<abcde>", False, True, True),
+        # FTP URL; not HTTP/HTTPS but has the necessary structure for
+        # completeness.
+        ("ftp://example.com/owner/resource", False, True, True),
+        # Contains an unencoded space, but still structured correctly.
+        ("https://example.com/owner/sub dir", False, False, False),
+        # Includes unencoded double quotes, but structured with necessary parts.
+        ('https://example.com/owner/"quotes"', False, False, False),
+    ],
+)
+def test_url_completeness_with_special_characters(
+    url, duplicate_flag, unsupported_url_scheme, expected_flag
+):
+    """Tests URL completeness check with special characters and schemes."""
+    df = pd.DataFrame(
+        {
+            "repourl": [url],
+            "duplicate_flag": [duplicate_flag],
+            "unsupported_url_scheme": [unsupported_url_scheme],
+        }
+    )
+    result = mark_incomplete_urls(df)
+    assert (
+        result["incomplete_url_flag"].iloc[0] == expected_flag
+    ), f"Completeness check for URL '{url}' failed."
+
+
+@pytest.mark.parametrize(
+    "data, expected_exception",
+    [
+        # Expect a ValueError for DataFrame without the 'repourl' column
+        ({"other_column": ["data1", "data2"]}, ValueError),
+    ],
+)
+def test_filter_incomplete_urls_exceptions(data, expected_exception):
+    """Verifies that the correct exception is raised for dataframes lacking
+    'repourl'."""
+    df = pd.DataFrame(data)
+    with pytest.raises(expected_exception):
+        mark_incomplete_urls(df)
+
+
+# Writing test cases for the function `extract_and_flag_domains`
+@pytest.mark.parametrize(
+    "data, expected_domains, expected_flags",
+    [
+        # Test with various supported and unsupported URL schemes
+        (
+            {
+                "repourl": [
+                    "https://github.com/openai",
+                    "http://example.com",
+                    "git://example.org/repo.git",
+                    "ftp://example.net/resource",
+                    "smb://fileserver/home",
+                ],
+                "duplicate_flag": [False, False, False, False, False],
+            },
+            [
+                pd.NA if domain is None else domain
+                for domain in ["github.com", "example.com", "example.org", None, None]
+            ],
+            [False, False, False, True, True],
```

Comment:
Again please could you add a short comment here to explain what this array contains (in other words its purpose).